include/coap3/coap_session.h

```diff
@@ -455,6 +464,36 @@ void coap_endpoint_set_default_mtu(coap_endpoint_t *endpoint, unsigned mtu);
  */
 COAP_API void coap_free_endpoint(coap_endpoint_t *endpoint);
 
+/**
+ * Returns any application-specific data that has been stored with @p
+ * endpoint using the function coap_endpoint_set_app_data2(). This function
+ * will return @c NULL if no data has been stored.
+ *
+ * @param endpoint The CoAP endpoint.
+ *
+ * @return Pointer to the stored data or @c NULL.
+ */
+void *coap_endpoint_get_app_data(const coap_endpoint_t *endpoint);
+
+/**
+ * Stores @p data with the given endpoint, returning the previously stored
+ * value or NULL. The data @p callback can be defined if the data is to be
+ * released when the endpoint is deleted.
+ *
+ * Note: It is the responsibility of the caller to free off (if appropriate) any
+ * returned data.
+ *
+ * @param endpoint The CoAP endpoint.
+ * @param data The pointer to the data to store or NULL to just clear out the
+ *             previous data.
+ * @param callback The optional release call-back for data on endpoint
+ *                 removal or NULL.
+ *
+ * @return The previous data (if any) stored in the endpoint.
+ */
+COAP_API void *coap_endpoint_set_app_data2(coap_endpoint_t *endpoint, void *data,
+                                           coap_app_data_free_callback_t callback);
```

Comment:
I would prefer this to be `coap_endpoint_set_app_data()` not `coap_endpoint_set_app_data2()` as the `2` suffix is usually used when a function needs to be upgraded with further functionality / parameter changes.

This needs changing in all the appropriate places.

I realize that you modelled this on `coap_session_set_app_data2()` (which is fine, but there is an older `coap_session_set_app_data()` with different parameters hence the 2 suffix here).